### PR TITLE
llama with HE

### PR DESCRIPTION
**Pull request by ciayomin**
fix some bugs when using HE on llama

### CODE REVIEW

**Comment by ciayomin** [comment]
@microsoft-github-policy-service agree

File: deepspeed/module_inject/containers/features/split_qkv.py [hunk 1]
```diff
@@ -33,21 +33,41 @@ def set_q_k_v(self):
...
+            self.module.attention.attn_ow[:, :self.dense_w.shape[1] // mp_replace.mp_size],
+            self.dense_w,
+            int8=reversed_dim,
+            allocate_tensor=reversed_dim) if self.dense_w is not None else None
+        self.module.attention.attn_ob = mp_replace.copy(
+            self.module.attention.attn_ob[:self.dense_w.shape[1] // mp_replace.mp_size],
+            self.dense_b,
+            int8=reversed_dim,
+            allocate_tensor=reversed_dim) if self.dense_b is not None else None
+
```
Comment (RezaYazdaniAminabadi) [inline deepspeed/module_inject/containers/features/split_qkv.py:70]: Thanks for solving this for the llama architecture. May I ask that you use the same style as suggested in the [PR](https://github.com/microsoft/DeepSpeed/pull/4077#pullrequestreview-1605864325), just to use less space to do the same changes. If most of these are already resolved in that PR, you can simply remove it as we will merge both this and other PR.

File: deepspeed/runtime/hybrid_engine.py [hunk 2]
```diff
@@ -227,6 +227,17 @@ def generate(self, *inputs, **kwargs):
...
+                attn_mask_shape = kwargs['attention_mask'].shape
+                output_attn_mask = torch.zeros(
+                    (attn_mask_shape[0] * self._config.hybrid_engine.inference_tp_size,) + attn_mask_shape[1:],
+                    dtype=kwargs['attention_mask'].dtype,
+                    device=kwargs['attention_mask'].device)
+                attn_mask_cont = kwargs['attention_mask'].contiguous()
+                dist.all_gather_into_tensor(output_attn_mask, attn_mask_cont, group=self.mp_group)
+
+                kwargs['attention_mask'] = output_attn_mask
+
```
Comment (RezaYazdaniAminabadi) [inline deepspeed/runtime/hybrid_engine.py:240]: This is a cool fix, thanks for fixing the attention mask when using MP.
